### PR TITLE
optimize object.assign({}, obj)

### PR DESCRIPTION
**Pull request by MikeHolman**
In case of simple object.assign({}, obj) we can memcpy the properties over to the new object.

### CODE REVIEW

File: lib/Runtime/Types/DynamicObject.cpp [hunk 1]
```diff
@@ -755,6 +755,94 @@ namespace Js
...
+        }
+        if (!from->GetTypeHandler()->AllPropertiesAreEnumerable())
+        {
+            if (PHASE_TRACE1(ObjectCopyPhase))
+            {
+                Output::Print(_u("ObjectCopy: Can't copy: from obj has non-enumerable properties\n"));
+            }
+            return false;
+        }
+        if (!from->GetTypeHandler()->IsPathTypeHandler())
```
Comment (agarwal-sandeep) [inline lib/Runtime/Types/DynamicObject.cpp]: >IsPathTypeHandler [](start = 37, length = 17)

nitpicking: Should this and ShareType check  be done first? Seems like these conditions are more frequent to fail?

Comment (MikeHolman) [reply lib/Runtime/Types/DynamicObject.cpp]: i'll move PathTypeHandler check up, but I put ShareType last since it has side effects in case it succeeds

File: lib/Runtime/Types/DynamicObject.cpp [hunk 2]
```diff
@@ -755,6 +755,94 @@ namespace Js
...
+
+        // Share the type
+        if (!from->GetDynamicType()->ShareType())
+        {
+            if (PHASE_TRACE1(ObjectCopyPhase))
+            {
+                Output::Print(_u("ObjectCopy: Can't copy: failed to share type\n"));
+            }
+            return false;
+        }
```
Comment (agarwal-sandeep) [inline lib/Runtime/Types/DynamicObject.cpp]: Can we add test cases for these ifs and boundary cases for aux and inline slot capacities?

File: lib/Runtime/Types/DynamicObject.cpp [hunk 3]
```diff
@@ -755,6 +755,94 @@ namespace Js
...
+        if (!from->GetTypeHandler()->IsPathTypeHandler())
+        {
+            if (PHASE_TRACE1(ObjectCopyPhase))
+            {
+                Output::Print(_u("ObjectCopy: Can't copy: Don't have PathTypeHandler\n"));
+            }
+            return false;
+        }
+        if(PathTypeHandlerBase::FromTypeHandler(from->GetTypeHandler())->HasAccessors())
```
Comment (agarwal-sandeep) [inline lib/Runtime/Types/DynamicObject.cpp]: >HasAccessors [](start = 73, length = 12)

Does this cover both getter and setter?

File: test/Object/rlexe.xml [hunk 4]
```diff
@@ -436,4 +436,10 @@
       <baseline />
     </default>
   </test>
+  <test>
+    <default>
+      <files>assign.js</files>
+      <compile-flags>-args summary -endargs</compile-flags>
```
Comment (agarwal-sandeep) [inline test/Object/rlexe.xml]: Shouldn't we have -trace:ObjectCopy so that we know the optimization is kicking in?

File: test/Object/assign.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    },
+    {
+        name: "has getter",
+        body: function ()
+        {
+            let orig = {};
+            orig.a = 1;
+            Object.defineProperty(orig, 'b', {
+                get: function() { return "asdf"; }, enumerable: true
+              });
```
Comment (agarwal-sandeep) [inline test/Object/assign.js]: Optimization should still kicking if prototype have getter? Can we add a test

File: test/Object/rlexe.xml (hunk 4 above)
Comment (MikeHolman) [reply test/Object/rlexe.xml]: I had that for my testing, but traces in tests can get noisy so didn't put it here. I can though since this probably shouldn't change much

File: lib/Runtime/Types/DynamicObject.cpp [hunk 6]
```diff
@@ -755,6 +755,94 @@ namespace Js
...
 
+    Field(Var)* DynamicObject::GetInlineSlots() const
+    {
+        return reinterpret_cast<Field(Var)*>(reinterpret_cast<size_t>(this) + this->GetOffsetOfInlineSlots());
+    }
+
+    bool DynamicObject::TryCopy(DynamicObject* from)
+    {
+        // Validate that objects are compatible
+        if (this->GetTypeHandler()->GetInlineSlotCapacity() != from->GetTypeHandler()->GetInlineSlotCapacity())
```
Comment (digitalinfinity) [inline lib/Runtime/Types/DynamicObject.cpp]: Is it worth extracting the compatibility checks to its own function? I'm wondering if it would be beneficial for the JIT to be able to generate a fast path for object.assign({}, obj) - in which case, as a simplification, it could call this function as a helper?

File: lib/Runtime/Types/DynamicObject.cpp [hunk 7]
```diff
@@ -755,6 +755,94 @@ namespace Js
...
+        if (!from->GetTypeHandler()->AllPropertiesAreEnumerable())
+        {
+            if (PHASE_TRACE1(ObjectCopyPhase))
+            {
+                Output::Print(_u("ObjectCopy: Can't copy: from obj has non-enumerable properties\n"));
+            }
+            return false;
+        }
+
+        // Share the type
```
Comment (digitalinfinity) [inline lib/Runtime/Types/DynamicObject.cpp]: Might be worth adding a comment here that ShareType has side effects so it should be done last

File: test/Object/assign.js [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    },
+    {
+        name: "proto accessor",
+        body: function ()
+        {
+            Object.defineProperty(Object.prototype, 'b', {
+                get: function() { return "asdf"; }
+              });
+            let orig = {};
+            orig.a = 1;
```
Comment (akroshg) [inline test/Object/assign.js]: Curious : symbol will not have impact, right?

File: lib/Runtime/Library/JavascriptObject.cpp [hunk 9]
```diff
@@ -1538,7 +1538,17 @@ namespace Js
             // else use enumerator to extract keys from source
             else
             {
-                AssignForGenericObjects(from, to, scriptContext);
+                DynamicObject* fromObj = JavascriptOperators::TryFromVar<DynamicObject>(from);
+                DynamicObject* toObj = JavascriptOperators::TryFromVar<DynamicObject>(to);
+                bool cloned = false;
```
Comment (akroshg) [inline lib/Runtime/Library/JavascriptObject.cpp]: say we failed to clone for the first source, do we still attempt to do for the next sources?

Comment (akroshg) [reply lib/Runtime/Library/JavascriptObject.cpp]: Do we need to check for cross-site object and just don't try that?

---
In reply to: [174883683](https://github.com/Microsoft/ChakraCore/pull/4817#discussion_r174883683) [](ancestors = 174883683)

Comment (akroshg) [reply lib/Runtime/Library/JavascriptObject.cpp]: Should we exclude external objects (IsExternal()). Not sure what is the harm though.

---
In reply to: [174883937](https://github.com/Microsoft/ChakraCore/pull/4817#discussion_r174883937) [](ancestors = 174883937,174883683)

Comment (MikeHolman) [reply lib/Runtime/Library/JavascriptObject.cpp]: I'll add case for cross-site object, I'm not sure about external, but I'll exclude them for good measure

File: test/Object/assign.js (hunk 8 above)
Comment (MikeHolman) [reply test/Object/assign.js]: nope, doesn't matter.  I added symbol to the test case
